### PR TITLE
Add tests for constructor checker

### LINKED ISSUE #350: Test constructor detection
- Consider edge cases

### PR DESCRIPTION
**Pull request by Nathan-SL**
## Usage related changes

-   NA
-   Closes #350 

## Development related changes

-   Introduces new test coverage for constructor chcker

## Checklist:

-   [x] Formatted the code
-   [x] No linter errors + tried to avoid introducing linter warnings
-   [x] Performed a self-review of the code
-   [x] Rebased to the last commit of the target branch (or merged it into my branch)
-   [ ] Documented the changes
-   [x] Updated the `test` directory (with a test case consisting of `network.json`, `hardhat.config.ts`, `check.ts`)
-   [x] Linked issues which this PR resolves
-   [x] Created a PR to the `plugin` branch of [`starknet-hardhat-example`](https://github.com/0xSpaceShard/starknet-hardhat-example):https://github.com/0xSpaceShard/starknet-hardhat-example/pull/113
    -   [x] Modified `test.sh` to use my example repo branch
    -   [x] Restored `test.sh` to to use the original branch (after the example repo PR has been merged)
-   [x] All tests are passing (for external contributors who don't have access to the CI/CD pipeline)

### CODE REVIEW

File: test/general-tests/constructor-test/check.ts [hunk 1]
```diff
@@ -1,4 +1,33 @@
...
+const contract2Path = path.join("cairo1-contracts", contract2);
+
+const contract3 = "mute_constructor.cairo";
+const contract3Path = path.join("cairo1-contracts", contract3);
+
+copyFileSync(path.join(prefix, contract1), contract1Path);
+copyFileSync(path.join(prefix, contract2), contract2Path);
+copyFileSync(path.join(prefix, contract3), contract3Path);
+
+const exptectedErrorMsg = "Error: Expected at most one constructor.";
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:24]:
```suggestion
const expectedErrorMsg = "Error: Expected at most one constructor.";
```

File: test/general-tests/constructor-test/check.ts [hunk 2]
```diff
@@ -1,4 +1,33 @@
...
+const contract1 = "duplicate_constructor.cairo";
+const contract1Path = path.join("cairo1-contracts", contract1);
+
+const contract2 = "no_constructor.cairo";
+const contract2Path = path.join("cairo1-contracts", contract2);
+
+const contract3 = "mute_constructor.cairo";
+const contract3Path = path.join("cairo1-contracts", contract3);
+
+copyFileSync(path.join(prefix, contract1), contract1Path);
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:20]: Add a comment before this line explaining that we are doing this to ensure file presence in the example repo

File: test/general-tests/constructor-test/check.ts [hunk 3]
```diff
@@ -1,4 +1,33 @@
-import { hardhatStarknetCompileDeprecated, hardhatStarknetTest } from "../../utils/cli-functions";
+import path from "path";
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:1]: I would put these new additions to a separate test folder and call it something like cairo1-constructor-test. Or maybe even better: change the name of this cairo0 test folder to cairo0-constructor-test (and call the cairo1 test just constructor-test)

File: test/general-tests/constructor-test/check.ts [hunk 4]
```diff
@@ -1,4 +1,33 @@
...
+import {
+    hardhatStarknetCompile,
+    hardhatStarknetCompileDeprecated,
+    hardhatStarknetTest
+} from "../../utils/cli-functions";
+import { copyFileSync } from "fs";
+import { assertContains, rmrfSync } from "../../utils/utils";
+
+const prefix = path.join(__dirname);
+const contract1 = "duplicate_constructor.cairo";
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:11]: Why not use descriptive variable naming? It could be `const duplicateConstructorContractName = "...`

File: test/general-tests/constructor-test/check.ts [hunk 5]
```diff
@@ -1,4 +1,33 @@
...
+const contract1Path = path.join("cairo1-contracts", contract1);
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:12]: `"cairo1-contracts"` should be extracted to a constant in this file called something like `sourcesPath`.

File: test/general-tests/constructor-test/duplicate_constructor.cairo [hunk 6]
```diff
@@ -0,0 +1,26 @@
...
+        balance::write(initial_balance);
+    }
+
+    #[constructor]
+    fn constructor2(initial_balance: felt252) {
+        balance::write(initial_balance);
+    }
+
+    #[external]
+    fn increase_balance(amount1: felt252, amount2: felt252) {
```
Comment (FabijanC) [inline test/general-tests/constructor-test/duplicate_constructor.cairo:18]: Here and in other new files, can we remove these unused external/view methods? Only constructors are needed, right? Maybe even remove the storage and initial_balance being a parameter of the constructor.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: test/general-tests/constructor-test/duplicate_constructor.cairo (hunk 6 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/duplicate_constructor.cairo]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/9b6048abf937d5c0936f1f5a4bf065acd4449935

File: test/general-tests/constructor-test/check.ts (hunk 5 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/check.ts]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/9b6048abf937d5c0936f1f5a4bf065acd4449935

File: test/general-tests/constructor-test/check.ts (hunk 4 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/check.ts]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/9b6048abf937d5c0936f1f5a4bf065acd4449935

File: test/general-tests/constructor-test/check.ts (hunk 3 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/check.ts:1]: Also did similar thing on the example repo. https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/9b6048abf937d5c0936f1f5a4bf065acd4449935

File: test/general-tests/constructor-test/check.ts (hunk 2 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/check.ts]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/9b6048abf937d5c0936f1f5a4bf065acd4449935

File: test/general-tests/constructor-test/check.ts [hunk 7]
```diff
@@ -1,8 +1,38 @@
...
-hardhatStarknetCompileDeprecated(
-    "contracts/contract.cairo contracts/simple_storage.cairo contracts/empty_constructor.cairo".split(
-        " "
-    )
-);
+const prefix = path.join(__dirname);
+const sourcesPath = "cairo1-contracts";
+
+const duplicateConstructorContract = "duplicate_constructor.cairo";
+const contract1Path = path.join(sourcesPath, duplicateConstructorContract);
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:10]: I think the descriptive naming should have been applied to the path variables as well. But I think we don't need the path variables really. We can use a loop instead of duplicating all the path joinings and copyFileSyncs. Maybe even the contract name variables aren't needed (except for this one because it has to be deleted later).

File: test/general-tests/constructor-test/check.ts [hunk 8]
```diff
@@ -1,8 +1,38 @@
...
+copyFileSync(path.join(prefix, commentedConstructorContract), contract4Path);
+copyFileSync(path.join(prefix, emptyLineConstructorContract), contract5Path);
+
+const expectedErrorMsg = "Error: Expected at most one constructor.";
+const execution = hardhatStarknetCompile([contract1Path], true);
+assertContains(execution.stderr, expectedErrorMsg);
+rmrfSync(contract1Path);
+
+// Compile cairo1 contracts
+hardhatStarknetCompile(["cairo1-contracts/", "--add-pythonic-hints"]);
```
Comment (FabijanC) [inline test/general-tests/constructor-test/check.ts:37]: Can't the sourcesPath variable be used here as well?

File: test/general-tests/constructor-test/no_constructor.cairo [hunk 9]
```diff
@@ -0,0 +1,12 @@
+#[contract]
+mod Contract {
+
+    struct Storage {
+        balance: felt252,
+    }
+
+    #[external]
+    fn increase_balance(amount1: felt252, amount2: felt252) {
```
Comment (FabijanC) [inline test/general-tests/constructor-test/no_constructor.cairo:9]: Did it not allow an empty constructor? If so, you can add a comment line documenting the purpose of this file and why it cannot be empty.

**Review by FabijanC — CHANGES_REQUESTED** [review]

File: test/general-tests/constructor-test/no_constructor.cairo (hunk 9 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/no_constructor.cairo]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/81d8dd0f8431b12eac70fcdd226edc426bae3432

File: test/general-tests/constructor-test/check.ts (hunk 8 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/check.ts]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/81d8dd0f8431b12eac70fcdd226edc426bae3432

File: test/general-tests/constructor-test/check.ts (hunk 7 above)
Comment (Nathan-SL) [reply test/general-tests/constructor-test/check.ts]: https://github.com/0xSpaceShard/starknet-hardhat-plugin/pull/377/commits/81d8dd0f8431b12eac70fcdd226edc426bae3432

**Review by FabijanC — APPROVED** [review]
